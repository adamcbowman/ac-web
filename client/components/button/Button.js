import React, { DOM } from 'react'
import PropTypes from 'prop-types'
import {
    compose,
    setPropTypes,
    setDisplayName,
    mapProps,
    defaultProps,
} from 'recompose'
import CSSModules from 'react-css-modules'
import styles from './Button.css'
import classNames from 'classnames'
import KIND, { ALL as KINDS } from './kinds'

const LEFT = 'LEFT'
const RIGHT = 'RIGHT'

export default compose(
    setDisplayName('Button'),
    setPropTypes({
        children: PropTypes.node,
        active: PropTypes.bool,
        shadow: PropTypes.bool,
        large: PropTypes.bool,
        transparent: PropTypes.bool,
        chevron: PropTypes.oneOf([LEFT, RIGHT, true]),
        kind: PropTypes.oneOf(Array.from(KINDS)),
        icon: PropTypes.node,
    }),
    defaultProps({
        kind: KIND,
        active: false,
        shadow: false,
        large: false,
        transparent: false,
    }),
    mapProps(
        ({
            icon,
            active,
            shadow,
            large,
            kind,
            transparent,
            chevron,
            ...props
        }) => {
            const styleName = classNames({
                [kind]: true,
                Active: active,
                Shadow: shadow,
                Large: large,
                Transparent: transparent,
                ChevronLeft: chevron === LEFT,
                ChevronRight: chevron === RIGHT,
                Chevron: chevron === true,
            })

            if (!icon) {
                return { ...props, styleName }
            }

            const { children } = props

            return {
                ...props,
                styleName: `${styleName} ${children ? '' : 'IconOnly'}`,
                children: (
                    <div styleName="IconContainer">
                        {icon}
                        {typeof children === 'string'
                            ? <span>{children}</span>
                            : children}
                    </div>
                ),
            }
        }
    ),
    CSSModules(styles, { allowMultiple: true })
)(DOM.button)
