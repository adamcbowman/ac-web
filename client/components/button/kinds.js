export const PRIMARY = 'Primary'
export const SECONDARY = 'Secondary'
export const TERTIARY = 'Tertiary'
export const INCOGNITO = 'Incognito'
export const SUBTILE = 'Subtile'

export const ALL = new Set([
    PRIMARY,
    SECONDARY,
    TERTIARY,
    INCOGNITO,
    SUBTILE,
])

export default PRIMARY
