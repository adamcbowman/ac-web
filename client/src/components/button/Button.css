@import "../../styles/components.css";

.Common {
    /*height: 2.5em;*/
    /*min-height: 30px;*/
    /*line-height: 2;*/
    padding: 0 12px;
    font-weight: 700;
    text-transform: uppercase;
	color: white;
	border: none;
    cursor: pointer;

    @media print {
        display: none;
    }
}

.Primary {
	composes: Common;

	background-color: var(--component-primary-bg);

	&:hover {
		background-color: var(--component-primary-bg-hover);
	}

	&:active, &.Active {
		background-color: var(--component-primary-bg-active);
	}
}

.Primary--Active {
    composes: Primary;

    background-color: var(--component-primary-bg-active);
}

.Secondary {
	composes: Common;

    background-color: var(--component-secondary-bg);

    &:hover {
        background-color: var(--component-secondary-bg-hover);
	}

    &:active, &.Active {
        background-color: var(--component-secondary-bg-active);
	}
}

.Tertiary {
	composes: Common;

    background-color: var(--component-tertiary-bg);

    &:hover {
        background-color: var(--component-tertiary-bg-hover);
	}

    &:active, &.Active {
        background-color: var(--component-tertiary-bg-active);
	}
}

.Incognito {
    composes: Common;

    background-color: transparent;
    color: var(--black);
}

.Subtile {
    composes: Common;

    background-color: transparent;
    color: var(--gray);

    & .IconContainer {
        & path:not([fill="none"]) {
            fill: var(--brand-primary);
        }
    }
}

.IconContainer {
    display: flex;
    align-items: center;

    & > :not(:first-child) {
        margin-left: 0.25em;
    }

    & .IconContainer {
        & path:not([fill="none"]) {
            fill: var(--gray);
        }
    }
}

.IconOnly {
    padding: 0 6px;
}

/* Modifiers */
.Large {
    padding: 1em;
    min-width: 5em;
    text-align: center;
}

.Chevron {
    &:after {
        content: " \203A";
        color: white;
        font-size: 1.25em;
    }
}
