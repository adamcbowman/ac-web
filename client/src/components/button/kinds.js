export const PRIMARY = 'Primary'
export const SECONDARY = 'Secondary'
export const TERTIARY = 'Tertiary'
export const INCOGNITO = 'Incognito'
export const SUBTILE = 'Subtile'

export default PRIMARY
