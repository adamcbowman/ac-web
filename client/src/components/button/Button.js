import React, { PropTypes, DOM} from 'react'
import {compose, setPropTypes, setDisplayName, mapProps, defaultProps} from 'recompose'
import CSSModules from 'react-css-modules'
import styles from './Button.css'
import * as KINDS from './kinds'
import {asValues} from 'constants/utils'

export default compose(
    setDisplayName('Button'),
    setPropTypes({
        children: PropTypes.node,
        active: PropTypes.bool,
        kind: PropTypes.oneOf(asValues(KINDS)),
        icon: PropTypes.node,
    }),
    defaultProps({
        kind: KINDS.default,
        active: false,
    }),
    mapProps(({icon, active, kind, ...props}) => {
        const styleName = `${kind} ${active ? 'Active' : ''}`

        if (!icon) {
            return {...props, styleName}
        }

        const {children} = props

        return {
            ...props,
            styleName: `${styleName} ${children ? '' : 'IconOnly'}`,
            children: (
                <div styleName='IconContainer'>
                    {icon}
                    {typeof children === 'string' ?
                        <span>{children}</span> :
                        children
                    }
                    </div>
            )
        }
    }),
    CSSModules(styles, { allowMultiple: true })
)(DOM.button)
