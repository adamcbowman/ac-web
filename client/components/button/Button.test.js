import React from 'react'
import Renderer from 'react-test-renderer'
import Button, { Close, Expand, Locate, Sorting } from './'
import { ALL } from './kinds'

function match(component) {
    expect(Renderer.create(component)).toMatchSnapshot()
}

test('button component', () => {
    match(<Button>Content</Button>)
    match(<Button active>Content</Button>)
    match(<Button shadow>Content</Button>)
    match(<Button large>Content</Button>)
    match(<Button transparent>Content</Button>)
    match(<Button chevron>Content</Button>)
    match(<Close>Content</Close>)
    match(<Expand>Content</Expand>)
    match(<Locate>Content</Locate>)
    match(<Sorting>Content</Sorting>)

    Array.from(ALL).forEach(kind => match(<Button kind={kind}>Content</Button>))
})
