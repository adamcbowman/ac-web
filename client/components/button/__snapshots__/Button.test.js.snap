// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button component 1`] = `
<button
  className="Primary"
  type="button"
>
  Content
</button>
`;

exports[`button component 2`] = `
<button
  className="Primary Active"
  type="button"
>
  Content
</button>
`;

exports[`button component 3`] = `
<button
  className="Primary Shadow"
  type="button"
>
  Content
</button>
`;

exports[`button component 4`] = `
<button
  className="Primary Large"
  type="button"
>
  Content
</button>
`;

exports[`button component 5`] = `
<button
  className="Primary Transparent"
  type="button"
>
  Content
</button>
`;

exports[`button component 6`] = `
<button
  className="Primary Chevron"
  type="button"
>
  Content
</button>
`;

exports[`button component 7`] = `
<button
  className="Close Subtile"
  type="button"
>
  Content
</button>
`;

exports[`button component 8`] = `
<button
  className="Subtile"
  type="button"
>
  <svg
    fill="none"
    height={24}
    viewBox="0 0 24 24"
    width={24}
  >
    <path
      d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
      fill="#245EAC"
    />
    <path
      d="M0 0h24v24H0z"
      fill="none"
    />
  </svg>
</button>
`;

exports[`button component 9`] = `
<button
  className="Subtile"
  type="button"
>
  <svg
    fill="none"
    height={24}
    viewBox="0 0 24 24"
    width={24}
  >
    <path
      d="M0 0h24v24H0z"
      fill="none"
    />
    <path
      d="M12 8c-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4zm8.94 3c-.46-4.17-3.77-7.48-7.94-7.94V1h-2v2.06C6.83 3.52 3.52 6.83 3.06 11H1v2h2.06c.46 4.17 3.77 7.48 7.94 7.94V23h2v-2.06c4.17-.46 7.48-3.77 7.94-7.94H23v-2h-2.06zM12 19c-3.87 0-7-3.13-7-7s3.13-7 7-7 7 3.13 7 7-3.13 7-7 7z"
      fill="#245EAC"
    />
  </svg>
</button>
`;

exports[`button component 10`] = `
<button
  className="Subtile"
  onClick={[Function]}
  title={null}
  type="button"
>
  <svg
    fill="none"
    height={24}
    viewBox="0 0 24 24"
    width={24}
  >
    <path
      d="M19 13H5v-2h14v2z"
      fill="#245EAC"
    />
    <path
      d="M0 0h24v24H0z"
      fill="none"
    />
  </svg>
</button>
`;

exports[`button component 11`] = `
<button
  className="Primary"
  type="button"
>
  Content
</button>
`;

exports[`button component 12`] = `
<button
  className="Secondary"
  type="button"
>
  Content
</button>
`;

exports[`button component 13`] = `
<button
  className="Tertiary"
  type="button"
>
  Content
</button>
`;

exports[`button component 14`] = `
<button
  className="Incognito"
  type="button"
>
  Content
</button>
`;

exports[`button component 15`] = `
<button
  className="Subtile"
  type="button"
>
  Content
</button>
`;
